[숫자 야구 게임] 김영은 미션 제출합니다.

wwwr-kim0en opened the pull request:
# 미션 - 숫자 야구
## 🚀 기능 요구 사항
- [X] : 사용자가 값을 입력할 수 있다.
  - [X] : 서로 다른 3자리 수
  - [X] :  재시작/종료 구분하는 1과2 중 하나의 수
- [X] : 컴퓨터가 랜덤하고 고유한 세 숫자를 정할 수 있다.
- [X] : 사용자의 값과 컴퓨터의 값을 비교하여 점수를 계산할 수 있다.
- [X] : 상황에 알맞은 문구 출력
  - [X] : 게임 시작 문구 :  숫자 야구 게임을 시작합니다.
  - [X] : 게임 재시작 경우, 시작 문구 제외
  - [X] : 점수 : 1볼 1스트라이크, 낫싱, 3스트라이크 등
  - [X] : 3스트라이크 시 종료 멘트 : 3개의 숫자를 모두 맞히셨습니다! 게임 종료
  - [X] : 점수 출력 후 종료 멘트 : 게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.
- [x] : 낫싱이 나오면 다시 사용자가 다시 숫자 입력하기(게임 종료 아님)


## 🎯프로그래밍 요구 사항 
- [X] : Node.js 18.17.1 버전에서 실행 가능하다.
- [X] : `play`메소드로 실행 가능하다.
- [X] : 외부 라이브러리 사용 안 한다.
- [X] : 순수 바닐라 JS로만 작성한다.
- [X] : [JavaScript 코드 컨벤션](https://github.com/woowacourse/woowacourse-docs/tree/main/styleguide/javascript)을 지키면서 개발
  - [X] : 소스의 변수명, 클래스명 등에는 영문 이외의 언어를 사용하지 않습니다.
  - [X] : 클래스, 메서드 등의 이름에는 특수 문자를 사용하지 않습니다.
  - [X] : 상수명은 SNAKE_CASE로 작성합니다.
- [X] : 프로그램 종료 시 `process.exit()`를 호출하지 않는다.
- [X] : 모든 테스트가 성공한다.
- [X] : 파일, 패키지 이름을 수정하거나 이동하지 않는다.

  
## ✏️ 과제 진행 요구 사항
- [X] : 미션은 [javascript-baseball](https://github.com/woowacourse-precourse/javascript-baseball-6/) 저장소를 Fork & Clone해 시작한다.
- [X] : **기능을 구현하기 전 `docs/README.md`에 구현할 기능 목록을 정리**해 추가한다.
- [X] : 과제 진행 및 제출 방법은 [프리코스 과제 제출](https://github.com/woowacourse/woowacourse-docs/tree/master/precourse) 문서를 참고한다.

Commit:
chore: 문제 요약해서 주석처리

Commit:
예외 상황들 정리

1. 입력값이 없을 때
2. 입력값이 숫자가 아닌 경우
3. 입력값(숫자)가 세자리가 아닌 경우
4. 입력값(숫자)가 중복될 때

Commit:
예외 상황들 코드로 작성

Commit:
config: 라이브러리 설치

@woowacourse/mission-utils

Commit:
1회차 풀이 - 오답

Commit:
computer Random 숫자 출력하는 함수 변경

Commit:
함수 생성- validateUserInput, compareUserComputer

Commit:
compareUserComputer 에서 스트라이크,볼 점수 반환

Commit:
수정: strike, ball 점수 계산

Commit:
게임 결과 반환 및 재시작/종료 묻기

Commit:
주석 제거

Commit:
fix: compareUserComputer 에서 return 위치 변경

Commit:
error 문구 변경

 [Error] 숫자가 잘못된 형식입니다.

Commit:
에러문 수정

Commit:
새로운 에러 상황 추가

숫자가 세자리이지만, 중복되는 숫자가 있을 경우

Commit:
숫자 중복 유무 확인하는 함수 생성

Commit:
trycatch문 async 있는 함수에 추가 & 에러 객체 수정

Commit:
n번째 풀이방식

Commit:
최종 풀이

Commit:
최최종 풀이

class MyError extends Error 생성

Commit:
예외성 검사 통과

Commit:
start() 와 end() 분리

Commit:
최최최종 풀이

Commit:
fix : 유효성 검사 통과

line 114 '[ERROR]' 추가

Commit:
style: 에러 안 나는 코드들 확인

Commit:
docs:  README.md 업데이트

Commit:
docs: README.md 업데이트

Commit:
docs: README.md 업데이트

프로그래밍 요구 사항 중 코드 컨벤션 상세 목록 추가

Commit:
style : 코드 컨벤션에 맞게 변수명 변명

Commit:
style: 함수 기능별로 쪼개기

Commit:
feat: 함수 exitApplication 작성

아예 어플리케이션 자체를 종료시키는 함수

Commit:
feat: 테스트 전부 통과

낫싱이 나오면 사용자가 재입력할 수 있게 돌아간다.

Commit:
style: 주석 삭제

Commit:
Merge branch 'www-r' of https://github.com/www-r/-javascript-baseball-6 into www-r

Commit:
docs: README.md 에 새로운 기능 목록에 추가하기

Commit:
docs: Update README.md

starcradle101 commented on src/App.js:
```diff
@@ -1,5 +1,129 @@
+import { MissionUtils } from '@woowacourse/mission-utils';
```
구조분해할당을 이용하면 필요한 라이브러리만 import 해서 더 명확하게 표현할 수 있을 것 같아요 😁

kim-young-dong replied:
동감합니다.

starcradle101 commented on src/App.js:
```diff
@@ -1,5 +1,129 @@
+	} catch (error) {}
+};
+
+const compareUserComputer = (USERARR, COMPUTERARR) => {
```
변수의 가독성이 떨어지는 것 같은데 언더바를 이용해서 작성하는 것은 어떨까요?

kim-young-dong replied:
개인적으로는 들여쓰기의 depth가 줄어들면 좋을것 같습니다.

starcradle101 commented on src/App.js:
```diff
@@ -1,5 +1,129 @@
+				throw new MyError('[ERROR]', '1부터 9까지의 자연수만 가능합니다.');
+			} else {
+				// USERARR 는 모두 자연수
+				if (USERARR[0] !== USERARR[1] && USERARR[0] !== USERARR[2] && USERARR[1] !== USERARR[2]) {
```
Set을 사용하면 중복 검사를 더욱 간편하게 할 수 있습니다 😆

migugin replied:
배열 중복 검사하는 방법으로 여러가지가 있는데 각각의 아이템을 하드 코딩 하는 것 보다 중복 검사 로직을 찾아보고 구현해보시면 좋을 것 같습니다!

starcradle101 commented on src/App.js:
```diff
@@ -1,5 +1,129 @@
+const getScore = (STRIKE, BALL, COMPUTERARR) => {
+	if (STRIKE === 3) {
+		MissionUtils.Console.print('3스트라이크');
+		MissionUtils.Console.print('3개의 숫자를 모두 맞히셨습니다! 게임 종료');
```
이스케이프 시퀀스를 사용하면 Console.print 한 번에 문장을 모두 표현할 수 있어요! 😌

kim-young-dong commented on README.md:
```diff
@@ -1,156 +1,34 @@
-
----
-
+- [X] : 사용자가 값을 입력할 수 있다.
```
요구사항을 꼼꼼하게 작성하셔서 좋습니다.

kim-young-dong reviewed (commented):
👍

Commit:
refactor: import 구조분해할당 , 상수 이름 변경 및 분리

1. {Random, Console}로 바꿔서 쓰는 애들만 import
2. 매직넘버/리터럴들은 따로 constants.js에 분리해서"SNAKE_CASE"로 작성

Commit:
refactor: constants.js 에 export 키워드 추가

Commit:
refactor: 매직넘버/리터럴 따로 분리 및 SNAKE_CASE로 변경

Commit:
chore: 안 쓰는 코드 정리

Commit:
refactor: 단항증가(++)/감소(--) 연산자 사용 지양하기

Commit:
refactor: else , else if 문 사용 지양하기.

return 을 하는 if문은 else 가 필요가 없다.

Commit:
refactor: iterable한 애들을 mapping 할 때는 Array.from 사용 지향 (보통은 spread)

Commit:
refactor: Object.freeze 를 통해 객체 불변성 유지하기

jinsoul75 commented on src/App.js:
```diff
@@ -1,5 +1,129 @@
+		MissionUtils.Console.print('3스트라이크');
+		MissionUtils.Console.print('3개의 숫자를 모두 맞히셨습니다! 게임 종료');
+		return;
+	} else if (STRIKE === 0 && BALL === 0) {
+		MissionUtils.Console.print('낫싱');
+		process(COMPUTERARR);
+		return;
+	} else if (STRIKE === 0) {
+		MissionUtils.Console.print(`${BALL}볼`);
+		return;
+	} else if (BALL === 0) {
+		MissionUtils.Console.print(`${STRIKE}스트라이크`);
+		return;
+	} else {
+		MissionUtils.Console.print(`${BALL}볼 ${STRIKE}스트라이크`);
+	}
```
else if 문은 가독성이 떨어지고 논리적 오류를 낼 수 있어 else문으로 early return을 사용하는것을 추천드립니다 ! :-)

kangju2000 commented on src/App.js:
```diff
@@ -1,5 +1,129 @@
+	}
+};
+
+const getScore = (STRIKE, BALL, COMPUTERARR) => {
```
`getScore` 메서드의 인자명이 모두 대문자로 이루어진 이유가 궁금해요 💭

wwwr-kim0en replied:
> `getScore` 메서드의 인자명이 모두 대문자로 이루어진 이유가 궁금해요 💭

얘는 제가 이유가 있었다기 보다는, 작업을 다 한 뒤에 상수들을 대문자로 바꾸다가 생긴 문제입니다 하하,, 리팩토링 때 얘도 수정하려고 합니다！ 지적 감사해요 ！！

helloworld442 reviewed (commented):
안녕하세요 코드 리뷰를 하게 된 minmin이라고 합니당!! 당차게 인사드립니다!!
먼저 정말 좋은 코드를 작성해주셔서 감사합니다. 배울 점이 많은 코드였습니다!!
특히 각 함수마다 예외상황을 작성해주시는 모습을 보고 정말 다양한 상황을 대비한다는 것이 좋았습니다.
다만 수정한다면 좀 더 좋을 거 같은 생각이 드는 데요!!
1. 변수명의 함축성
2. 예외 상황에 대한 에러 처리

3. depth 설정

먼저 첫번째 변수명의 함축성 부분인데요!! 변수를 보고 아 이건 이런 걸 하는 함수 이구나 라는 거를 인지할 수 있는 코드도 있었지만 반대로 이건 어떤 작업을 하는 코드지? 라는 생각이 드는 부분도 있었습니다. (ex ) start , process , end) 이런 부분들에 대해서 조금더 명확하게 의미를 전달해주시면 바로 이 코드가 어떤 코드 인지 알 수 있을 거 같습니다.
두번째는 모든 상황에 대한 예외 처리가 제대로 되고 있지 않는 것 같아용!! try 와 catch 구문을 사용한 throw new Error를 처리 하는 부분은 좋았지만 그렇지 않고 catch 구문에서 예외 처리가 되고 있지 않는 부분도 있는 것 같아요!! 나중에 확인하시고 이에 대해서 생각해 주시면 좋을 거 같습니다.
세번째는 depth 설정입니당!! depth설정은 제대로 되고 있지만 들어쓰기의 간격이 다소 길다는 생각이 듭니다.

kangju2000 commented on src/App.js:
```diff
@@ -1,5 +1,129 @@
+	await process(COMPUTERARR);
+};
+
+const process = async (COMPUTERARR) => {
```
`process`를 사용하는 코드를 보고 Node.js의 `process`가 생각이 났어요 ㅎㅎ

통용되는 이름을 사용하기보다는, 다른 함수명으로 바꾸는건 어떨까용?

wwwr-kim0en replied:
> `process`를 사용하는 코드를 보고 Node.js의 `process`가 생각이 났어요 ㅎㅎ
> 
> 통용되는 이름을 사용하기보다는, 다른 함수명으로 바꾸는건 어떨까용?

Node.js 는 아직 공부를 해보지 않아서 몰랐던 부분이네요！ 감사합니다 ㅎㅎ

kangju2000 commented on src/App.js:
메시지를 출력하거나, 게임 룰(입력 개수, 입력 범위 등..)을 `상수`로 분리해서 사용하면 나중에 변경될 때 하나하나 고치지 않고 한번에 고칠 수 있어 유지보수하기 편해질거라고 생각해요!

migugin commented on src/App.js:
```diff
@@ -1,5 +1,129 @@
+	}
+};
+
+const getScore = (STRIKE, BALL, COMPUTERARR) => {
+	if (STRIKE === 3) {
+		MissionUtils.Console.print('3스트라이크');
+		MissionUtils.Console.print('3개의 숫자를 모두 맞히셨습니다! 게임 종료');
+		return;
+	} else if (STRIKE === 0 && BALL === 0) {
+		MissionUtils.Console.print('낫싱');
+		process(COMPUTERARR);
+		return;
+	} else if (STRIKE === 0) {
+		MissionUtils.Console.print(`${BALL}볼`);
+		return;
+	} else if (BALL === 0) {
+		MissionUtils.Console.print(`${STRIKE}스트라이크`);
+		return;
+	} else {
+		MissionUtils.Console.print(`${BALL}볼 ${STRIKE}스트라이크`);
+	}
+};
```
else if 는 가독성/성능 측면에서 모두 지양하는 편이 좋습니다.
만약 조건문의 else 문을 실행하게 되면 위에 작성된 else if 문의 조건을 모두 판별한 뒤에 작동하여 성능이 저하되기 때문입니다!

wwwr-kim0en replied:
> else if 는 가독성/성능 측면에서 모두 지양하는 편이 좋습니다. 만약 조건문의 else 문을 실행하게 되면 위에 작성된 else if 문의 조건을 모두 판별한 뒤에 작동하여 성능이 저하되기 때문입니다!

이유까지 같이 설명해주셔서 감사합니다 ㅎ！

kangju2000 commented on src/App.js:
```diff
@@ -1,5 +1,129 @@
+
+const getComputerInput = () => {
+	try {
+		const COMPUTERARR = [];
```
자바스크립트에서는 변수명으로 `camelCase`를 사용하는 것을 권장하고 있어요.

그래서 `COMPUTERARR`이라는 변수보다는, `computerArr` 또는 `computerArray` 처럼 두는게 좋아보여요 😄

kangju2000 reviewed (commented):
1주차 과제 고생하셨어요 🎉🎉
2주차도 화이팅!!

migugin commented on src/App.js:
```diff
@@ -1,5 +1,129 @@
+	await process(COMPUTERARR);
+};
+
+const process = async (COMPUTERARR) => {
```
상수명을 입력할 때 모두 대문자로 쓰셨는데 아마 요구 사항을 지키려고 하신 것 같습니다!
다만 모두 대문자로 썼을 때 이 상수가 무엇을 의미하는지 파악하기 어렵기 때문에 언더바로 구분해주시면 좋을 듯 합니다.

Computer Array / User Input 등의 변수는 항상 고정되어 있는 값이 아닌 게임이 실행될 때마다 변화하는 값이기 때문에,
computerArr / userArr 등의 이름으로 만들어 주시면 좋을 것 같습니다.

migugin reviewed (commented):
README 를 굉장히 자세하게 작성해주셨네요!
그 외 제가 말씀드린 부분을 다들 이미 피드백을 주신 것 같긴 하네요 😂
1주차 피드백 적용하셔서 2주차 미션도 같이 화이팅 합시다!

Commit:
refactor: 함수명 변경

process 는 Node.js 에 존재하는 함수 이름이기 때문에 다른 이름으로 변경

Commit:
refactor: 함수명 변경

start, end => startGame, endGame
